host/class/uac/usb_host_uac/test_app/main/test_host_uac.c

```diff
@@ -0,0 +1,904 @@
+    TEST_ASSERT_NOT_NULL(tx_buffer);
+    uint32_t tx_size = tx_buffer_threshold;
+
+    int freq_offsite_step = sample_freq / spk_alt_params.sample_freq[0];
+    int downsampling_bits = bit_resolution - spk_alt_params.bit_resolution;
+    int offset_size = tx_size / (spk_alt_params.bit_resolution / 8);
+    // we can only support adjust from 2 channels to 1 channel
+    bool channels_adjust = channels != spk_alt_params.channels ? true : false;
+
+    // fill the tx buffer with wav file data
+    for (size_t i = 0; i < offset_size; i++) {
+        tx_buffer[i] = s_buffer[i * freq_offsite_step] >> downsampling_bits;
+    }
+
+    if (downsampling_bits == 8) {
+        // move buffer to the correct position
+        for (size_t i = 0; i < offset_size; i++) {
+            *((uint8_t *)tx_buffer + i) = *((uint8_t *)tx_buffer + i * 2);
+        }
+        tx_size = tx_size / 2;
+    } else if (downsampling_bits) {
+        ESP_LOGE(TAG, "Unsupported downsampling bits %d", downsampling_bits);
+    }
+
+    if (channels_adjust) {
+        // convert stereo to mono
+        if (downsampling_bits == 8) {
+            tx_size = tx_size / 2;
+            for (size_t i = 0; i < tx_size; i++) {
+                *((uint8_t *)tx_buffer + i) = *((uint8_t *)tx_buffer + i * 2);
+            }
+        } else {
+            tx_size = tx_size / 2;
+            for (size_t i = 0; i < tx_size; i++) {
+                tx_buffer[i] = tx_buffer[2 * i];
+            }
+        }
+    }
+    s_buffer += offset_size * freq_offsite_step;
+
+    // invalid operate
+    uint32_t volume = 10;
+    TEST_ASSERT_EQUAL(ESP_ERR_INVALID_STATE, uac_host_device_write(uac_device_handle, (uint8_t *)tx_buffer, tx_size, 0));
+    TEST_ASSERT_EQUAL(ESP_OK, uac_host_device_set_mute(uac_device_handle, 0));
+    TEST_ASSERT_EQUAL(ESP_OK, uac_host_device_set_volume(uac_device_handle, volume));
+    TEST_ASSERT_EQUAL(ESP_OK, uac_host_device_resume(uac_device_handle));
+    TEST_ASSERT_EQUAL(ESP_OK, uac_host_device_write(uac_device_handle, (uint8_t *)tx_buffer, tx_size, 0));
+
+    // playback from volume 10 to 100
+    const uint32_t volume_max = 100;
+    event_queue_t evt_queue = {0};
+    while (1) {
+        if (xQueueReceive(s_event_queue, &evt_queue, portMAX_DELAY)) {
+            TEST_ASSERT_EQUAL(UAC_DEVICE_EVENT, evt_queue.event_group);
+            uac_host_device_handle_t uac_device_handle = evt_queue.device_evt.handle;
+            uac_host_device_event_t event = evt_queue.device_evt.event;
+            switch (event) {
+            case UAC_HOST_DEVICE_EVENT_TX_DONE:
+                if ((uint32_t)(s_buffer + offset_size) > (uint32_t)wav_file_end) {
+                    s_buffer = (uint16_t *)(wav_file_start + 44);
+                    volume += 20;
+                    TEST_ASSERT_EQUAL(ESP_OK, uac_host_device_set_volume(uac_device_handle, volume > 100 ? 100 : volume));
+                    printf("Volume: %" PRIu32 "\n", volume);
```

Comment:
I tired running the test app, and the volume was set from 10 to 110 on my headphones. 
Is there any reason for that? 

In a comment above (line 541) you are stating that the volume will be set from 10 to 100, which of course will, because of the `uac_device_handle, volume > 100 ? 100 : volume` in the line above. 
But this printf will say the volume is 110.

Reply:
the printf should also be `printf("Volume: %" PRIu32 "\n", volume > 100 ? 100 : volume);`, fixed.